modules/tracktion_engine/timestretch/tracktion_RubberBand.cpp

```diff
@@ -0,0 +1,269 @@
+
+        // Check number of zero crossings and estimate pitch
+        int numZeroCrossingsShifted = getNumZeroCrossings (sinBuffer);
+        float shiftedPitch = getPitchFromNumZeroCrossings (numZeroCrossingsShifted, sinBuffer.getNumSamples(), 44100);
+
+        // Compare shiftedPitch to originalPitch with a tolerance
+        expectLessThan (std::abs (shiftedPitch - (457.69f)), 2.0f);
+    }
+
+    void runTimeStretchTest()
+    {
+        beginTest ("Time-stretch");
+
+        RubberStretcher thisRubberBandStretcher = RubberStretcher (44100, 512, 2, RubberBandStretcher::Option::OptionProcessRealTime);
+
+        thisRubberBandStretcher.reset();
+
+        // Stretches the audio by /2, at the same pitch
+        thisRubberBandStretcher.setSpeedAndPitch (0.5f, 1.0f);
+
+        AudioBuffer<float> sinBuffer;
+        sinBuffer.setSize (2, 44100);
+
+        auto wL = sinBuffer.getWritePointer (0);
+        auto wR = sinBuffer.getWritePointer (1);
+
+        double currentAngle = 0.0, angleDelta = 0.0;
+        float originalPitch = 432.0f; //A4
+        auto cyclesPerSample = originalPitch / 44100.0f;
+        angleDelta = cyclesPerSample * 2.0 * juce::MathConstants<double>::pi;
+
+        // Generate sin wave and store in testBuffer
+        for (int sample = 0; sample < 44100; ++sample)
+        {
+            auto currentSample = (float) std::sin (currentAngle);
+            currentAngle += angleDelta;
+            wL[sample] = currentSample;
+            wR[sample] = currentSample;
+        }
+
+        // Process sin wave to time stretch and store in testBuffer
+        thisRubberBandStretcher.processData (sinBuffer.getArrayOfReadPointers(), 44100, sinBuffer.getArrayOfWritePointers());
+
+        AudioBuffer<float> testBuffer;
+        testBuffer.setSize (2, 22050);
+
+        auto t = testBuffer.getWritePointer (0);
+        auto s = sinBuffer.getWritePointer (0);
+
+        int scount = 0;
+
+        for (int sample = 0; sample < 44100; ++sample)
+        {
+            if (s[sample] != 0 && sample >= 22050)
+            {
+                t[sample] = s[sample];
+            }
+            else
+            {
+                scount++;
+            }
+        }
+
+        // Check whether buffer has resized
+        expectEquals (scount, 22050);
+        expectLessThan (std::abs (sinBuffer.getNumSamples() - testBuffer.getNumSamples() * 2), 10);
+    }
+
+    //==============================================================================
+    float getPitchFromNumZeroCrossings (int numZeroCrossings, int numSamples, double sampleRate)
+    {
+        float bufferTimeInSeconds = numSamples / sampleRate;
+
+        int numCycles = numZeroCrossings / 2;
+
+        float pitchInHertz = numCycles / bufferTimeInSeconds;
+
+        return pitchInHertz;
+    }
+
+    int getNumZeroCrossings (AudioBuffer<float> buffer)
```

Comment:
This `AudioBuffer<float> buffer` is taking a copy of the buffer which is unnecessarily expensive. Because this buffer isn't mutated in any way by the function it should be a const reference.